// mattermost-redux/store/configureStore.dev.js
"use strict";

require("core-js/modules/web.dom-collections.iterator");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = configureServiceStore;

var _redux = require("redux");

var _remoteReduxDevtools = _interopRequireDefault(require("remote-redux-devtools"));

var _reduxOffline = require("redux-offline");

var _defaults = _interopRequireDefault(require("redux-offline/lib/defaults"));

var _reducer_registry = _interopRequireDefault(require("./reducer_registry"));

var _reducers = _interopRequireDefault(require("../reducers"));

var _deep_freeze = _interopRequireDefault(require("../utils/deep_freeze"));

var _initial_state = _interopRequireDefault(require("./initial_state"));

var _helpers = require("./helpers");

var _middleware = require("./middleware");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

/* eslint-disable no-undefined */
const devToolsEnhancer = typeof window !== 'undefined' && window.__REDUX_DEVTOOLS_EXTENSION__ ? // eslint-disable-line no-underscore-dangle
window.__REDUX_DEVTOOLS_EXTENSION__ : // eslint-disable-line no-underscore-dangle
() => {
  return (0, _remoteReduxDevtools.default)({
    name: 'Mattermost',
    hostname: 'localhost',
    port: 5678,
    realtime: true
  });
};

/**
 * Configures and constructs the redux store. Accepts the following parameters:
 * preloadedState - Any preloaded state to be applied to the store after it is initially configured.
 * appReducer - An object containing any app-specific reducer functions that the client needs.
 * userOfflineConfig - Any additional configuration data to be passed into redux-offline aside from the default values.
 * getAppReducer - A function that returns the appReducer as defined above. Only used in development to enable hot reloading.
 * clientOptions - An object containing additional options used when configuring the redux store. The following options are available:
 *     additionalMiddleware - func | array - Allows for single or multiple additional middleware functions to be passed in from the client side.
 *     enableBuffer - bool - default = true - If true, the store will buffer all actions until offline state rehydration occurs.
 *     enableThunk - bool - default = true - If true, include the thunk middleware automatically. If false, thunk must be provided as part of additionalMiddleware.
 */
function configureServiceStore(preloadedState, appReducer, userOfflineConfig, getAppReducer, clientOptions) {
  const baseOfflineConfig = Object.assign({}, _defaults.default, _helpers.offlineConfig, userOfflineConfig);
  const baseState = Object.assign({}, _initial_state.default, preloadedState);
  const loadReduxDevtools = process.env.NODE_ENV !== 'test'; //eslint-disable-line no-process-env

  const store = (0, _redux.createStore)((0, _reduxOffline.createOfflineReducer)(createDevReducer(baseState, _reducers.default, appReducer)), baseState, (0, _reduxOffline.offlineCompose)(baseOfflineConfig)((0, _middleware.createMiddleware)(clientOptions), loadReduxDevtools ? [devToolsEnhancer()] : []));

  _reducer_registry.default.setChangeListener(reducers => {
    store.replaceReducer((0, _reduxOffline.createOfflineReducer)(createDevReducer(baseState, reducers)));
  }); // launch store persistor


  if (baseOfflineConfig.persist) {
    baseOfflineConfig.persist(store, baseOfflineConfig.persistOptions, baseOfflineConfig.persistCallback);
  }

  if (baseOfflineConfig.detectNetwork) {
    baseOfflineConfig.detectNetwork(online => {
      store.dispatch((0, _reduxOffline.networkStatusChangedAction)(online));
    });
  }

  if (module.hot) {
    // Enable Webpack hot module replacement for reducers
    module.hot.accept(() => {
      const nextServiceReducer = require("../reducers").default; // eslint-disable-line global-require


      let nextAppReducer;

      if (getAppReducer) {
        nextAppReducer = getAppReducer(); // eslint-disable-line global-require
      }

      store.replaceReducer(createDevReducer(baseState, _reducer_registry.default.getReducers(), nextServiceReducer, nextAppReducer));
    });
  }

  return store;
}

function createDevReducer(baseState, ...reducers) {
  return enableFreezing((0, _helpers.createReducer)(baseState, ...reducers));
}

function enableFreezing(reducer) {
  return (state, action) => {
    const nextState = reducer(state, action);

    if (nextState !== state) {
      (0, _deep_freeze.default)(nextState);
    }

    return nextState;
  };
}
