// mattermost-redux/actions/alerts.js
"use strict";

require("core-js/modules/es.promise");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.pushNotificationAlert = pushNotificationAlert;
exports.pushDeveloperAlert = pushDeveloperAlert;
exports.pushErrorAlert = pushErrorAlert;
exports.clearLatestAlert = clearLatestAlert;

var _action_types = require("../action_types");

var _constants = require("../constants");

// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
// @flow
function pushNotificationAlert(message
/*: string*/
)
/*: ActionFunc*/
{
  return async (dispatch, getState) => {
    const notificationAlert
    /*: AlertType*/
    = {
      type: _constants.Alerts.ALERT_NOTIFICATION,
      message
    };
    dispatch({
      type: _action_types.AlertTypes.PUSH_ALERT,
      data: notificationAlert
    }, getState);
    return {
      data: true
    };
  };
}

function pushDeveloperAlert(message
/*: string*/
)
/*: ActionFunc*/
{
  return async (dispatch, getState) => {
    const developerAlert
    /*: AlertType*/
    = {
      type: _constants.Alerts.ALERT_DEVELOPER,
      message
    };
    dispatch({
      type: _action_types.AlertTypes.PUSH_ALERT,
      data: developerAlert
    }, getState);
    return {
      data: true
    };
  };
}

function pushErrorAlert(message
/*: string*/
)
/*: ActionFunc*/
{
  return async (dispatch, getState) => {
    const errorAlert
    /*: AlertType*/
    = {
      type: _constants.Alerts.ALERT_ERROR,
      message
    };
    dispatch({
      type: _action_types.AlertTypes.PUSH_ALERT,
      data: errorAlert
    }, getState);
    return {
      data: true
    };
  };
}

function clearLatestAlert()
/*: ActionFunc*/
{
  return async (dispatch, getState) => {
    dispatch({
      type: _action_types.AlertTypes.CLEAR_ALERT,
      data: null
    }, getState);
    return {
      data: true
    };
  };
}
